exp/views/responses.py

```diff
@@ -1108,6 +1160,62 @@ def get(self, request, *args, **kwargs):
         set_content_disposition(response, filename)
         return response
 
+class StudyResponsesFrameDataPsychDS(ResponseDownloadMixin, generic.list.ListView):
+    """Hitting this URL downloads a ZIP file in Psych-DS formatting with frame data from one response per file in CSV format"""
+
+    # TODO: with large files / many responses generation can take a while. Should generate asynchronously along
+    # with the data dict.
+    def render_to_response(self, context, **response_kwargs):
+        paginator = context["paginator"]
+        study = self.study
+
+        if study.study_type.is_external:
+            messages.error(
+                self.request, "Frame data is not available for External Studies."
+            )
+
+            return study_responses_all(study)
+        
+        # build dict with contextual infomation about the study/lab/user
+        metadata_json = build_metadata_object(study)
+        # variables from column headers will be collected while looping through files
+        variables_measured = []
+
+        zipped_file = io.BytesIO()
+        with zipfile.ZipFile(zipped_file, "w", zipfile.ZIP_DEFLATED) as zipped:
+            for page_num in paginator.page_range:
+                page_of_responses = paginator.page(page_num)
+                for resp in page_of_responses:
+                    data = build_single_response_framedata_csv(resp)
+                    # get column headers from first row
+                    variables_measured += [column.strip('"') for column in data.split('\n')[0].strip().split(',')]
+                    # psych-DS files use "keyword" formatting
+                    keywords = {'study':study_name_for_files(study.name),'response':resp.id,'child':resp.child_id}
+                    filename = keyword_filename(keywords,"csv")
+                    zipped.writestr(f"/data/{filename}", data)
+                    #make a response-specific metadata file
+                    sidecar_metadata = {
+                        "response_uuid":str(resp.uuid),
+                        "eligibility":resp.eligibility[0],
+                        "study_completed":resp.completed
+                    }
+                    sidecar_filename = keyword_filename(keywords,"json")
+                    zipped.writestr(f"/data/{sidecar_filename}", json.dumps(sidecar_metadata,indent=4))
+            zipped.writestr(f"/materials/study_protocol.json", json.dumps(study.structure,indent=4))
+            #add final variables list to metadata
+            metadata_json['variableMeasured'] = variables_measured
+            zipped.writestr('dataset_description.json',f"{json.dumps(metadata_json,indent=4)}")
+
+        zipped_file.seek(0)
+        response = FileResponse(
+            zipped_file,
```

Comment:
Why seek back to the start of the file?  Why not exit the `with` block effectively closing the file and return that in the response?

Reply:
Not to say I think this is wrong, I am just curious if you have a reason why this would be better.

Reply:
I can't really explain why this is necessary, but for some reason the download stalls out when I try to remove it. I agree that the with block should make this unnecessary.